### ISSUE TITLE
Call for API review

### ISSUE THREAD

**Issue by gndl** [savonet/ocaml-ffmpeg#11]
Before considering a release, if you have a little time, I would like to have your point of view on the API which you can consult the documentation [here](http://liquidsoap.fm/ocaml-ffmpeg/docs/api/index.html). I would also like to make this request on [discuss.ocaml.org](https://discuss.ocaml.org) if you do not see any inconvenience.
(My apologies if the issues tracker is not the right place for this request.)
Thank you in advance.

**Comment by toots** [comment]
Awesome. I'll try to have a look at it next week. From what I've seen it looked like you did a great job with it.

**Comment by toots** [comment]
I'm also adding @smimram to the issue, maybe he'll have a minute to look at it :-)

**Comment by toots** [comment]
Yeah, it looks great. I'm gonna work on integrating it with liquidsoap, that should give me a better understanding of the API.

One question, tho: did you consider using polymorphic variant types for all the declarative types like codecs, etc? That would potentially make it easier I think to adapt to changing list of codecs. You could even auto-generate the list of actually compiled codec and have something like:

```
type all_possible_codecs = [`Mp3 | ... ]
val available_codecs : unit -> [< all_possible_codecs]
```

**Comment by gndl** [comment]
> did you consider using polymorphic variant types
No. Do you have a pointer to an example of code?

**Comment by toots** [comment]
Ok, I have started implementing the new API in liquidsoap. As expected, I really like the API, it is self-explanatory and has a clear and clean workflow. PR is here: https://github.com/savonet/liquidsoap/pull/510

Couple of remarks:
* As I was saying earlier, with APIs driven by type variants to describe data, using polymorphic variants removes some of the pain, for instance in the PR, we have:
```
      match FFmpeg.Av.read stream with
           | FFmpeg.Av.Frame frame -> convert frame
          | FFmpeg.Av.End_of_stream -> [|[||];[||]|]
     in
```
This could be replaced by:
```
      match FFmpeg.Av.read stream with
           | `Frame frame -> convert frame
           |  `End_of_stream -> [|[||];[||]|]
     in
```

* It seems that there's some clipping in the decoded float data coming back from:
```
module ConverterInput = FFmpeg.Swresample.Make(FFmpeg.Swresample.Frame)
module Converter = ConverterInput(FFmpeg.Swresample.PlanarFloatArray)
```
Should this data by in interval `(-1.,1.)`? Otherwise, could the doc specify (and possibly provide an efficient C-based normalization to `(-1.,1.)`?)

Thanks for the great work! I'll keep adding decoder and etc to liquidsoap and will report more.

**Comment by toots** [comment]
Another issue I am having: it seems that calling `get_input_duration` right after `open_input` does not work while it works when called on a specific stream:
```
  let container =
    FFmpeg.Av.open_input file
  in
  let duration =
    FFmpeg.Av.get_input_duration container FFmpeg.Avutil.Time_format.Millisecond
  in
```
Results in error: `FFmpeg.Avutil.Failure("Failed to get stream -1 duration : index out of bounds")`

While this works:
```
  let container =
    FFmpeg.Av.open_input file
  in
  let (index, stream, codec) =
    FFmpeg.Av.find_best_audio_stream container
  in
  let duration =
    FFmpeg.Av.get_duration stream FFmpeg.Avutil.Time_format.Millisecond
  in
```

**Comment by gndl** [comment]
> This could be replaced by:

>      match FFmpeg.Av.read stream with
>        | `Frame frame -> convert frame
>        | `End_of_stream -> [|[||];[||]|]

Indeed, this is nicer. I can do this improvement.
I started working on auto-generating the codec ids. I should be able to do this improvement for these too.
The other two observations seem to highlight some bugs. I will look at it.

**Comment by gndl** [comment]
The Av.get_input_duration fix is committed.
I can't reproduce the interval problem. Which FFmpeg version do you use? Can you send me the audio file with which you have the problem?

**Comment by toots** [comment]
Nice, thanks for that. I'll test and report ASAP.

**Comment by toots** [comment]
I've been looking at integrating the API further into liquidsoap, here are more questions for you:
* How do you decode video streams? I'm not sure how to access video frames. Do you have an API to decode them as big array?
* Same questions with encoding video :-)
* Can you encode data as a pipe, not just to a file? I see two calls to create output containers from format ID or name but I'm not sure how to access encoded data.
* Could it be possible to add an API to decode as a pipe,  not juste from a file? That would allow us to use the decoder for streams.

By "as a pipe" I mean as an API which receives and emits bytes.

Thanks for you hard work on all this!

**Comment by toots** [comment]
Sorry I forgot to detail the clipping issue. I was using the `FFMPEG` decoder from the liquidsoap PR above. Any chance you could test it from that liquidsoap branch? I could provide instructions to reproduce then..

**Comment by gndl** [comment]
> How do you decode video streams? I'm not sure how to access video frames. Do you have an API to decode them as big array?

I am not really concerned by the video part so i didn't do the video frame/planes (bigarray) conversion functions.
You could use the Avutil.Video.frame_get function to read the decoded data but it would probably be slow and certainly unsafe because i didn't provide the API to get the lines size.
You can use the Avutil.Video.frame_set function to produce video frames like [here](https://github.com/savonet/ocaml-ffmpeg/blob/master/examples/encode_video/encode_video.ml) but with the same issues.
For transcoding, the Av.write function automaticaly scale the video frames (and resample the audio frames) according to the output format if needed.

> Can you encode data as a pipe

Not for now but it could be feasible.

> Could it be possible to add an API to decode as a pipe

Yes, as well.

> Any chance you could test it from that liquidsoap branch?

I will try with the help of your instructions.

**Comment by gndl** [comment]
I committed the codec identifiers and the time format as polymorphic variant types. 
Now, it would be logical to do the same for channel layout and **pixel format** but it would break the API compatibility.
Would that be acceptable?

**Comment by toots** [comment]
Absolutely! It's remarkable you were able to keep it backward compatible so far but yeah, don't hesitate to break it, I'll update our code.

**Comment by toots** [comment]
> I am not really concerned by the video part so i didn't do the video frame/planes (bigarray) conversion functions.
> You could use the Avutil.Video.frame_get function to read the decoded data but it would probably be slow and certainly unsafe because i didn't provide the API to get the lines size.
> You can use the Avutil.Video.frame_set function to produce video frames like here but with the same issues.
> For transcoding, the Av.write function automaticaly scale the video frames (and resample the audio frames) according to the output format if needed.

Ok. Feel free to take your time on this work. For us, video decoding/encoding using `ffmpeg` is where the good stuff would be. We've got pretty much all bases covered for audio with the different bindings but that's not the case at all with video, which is much more complicated to handle. There's gstreamer that does a good job but its code has complicated issues so being able to use `ffmpeg` would be awesome.

>> Can you encode data as a pipe
> Not for now but it could be feasible.
>
>> Could it be possible to add an API to decode as a pipe
> Yes, as well.

Same thing here, take your time but let me know if you ever implement it, I'd be very happy to use it in liquidsoap!

>> Any chance you could test it from that liquidsoap branch?
> I will try with the help of your instructions.

Ok, here we go:
```
% cd /path/to/ffmpeg
% opam pin add ffmpeg .
(installs ffmpeg.. )
% cd /some/other/path
% git clone https://github.com/savonet/liquidsoap.git
% cd liquidsoap
% opam pin add liquidsoap .
```

You should end-up with a liquidsoap compiled using the latest `ffmpeg`. It is sometimes more convenient to directly link to locally built `ffmpeg`, in this case you would do:
```
(inside liquidsoap directory)
% ./configure --with-ffmpeg-dir=/path/to/ocaml-ffmpeg/src
% make
(in this case, use the binary at ./src/liquidsoap)
```

Once you have  it all compiled you can for instance output to a `wav` file:
```
% liquidsoap 'output.file(%wav,"/tmp/test.wav",single("/path/to/file.mp3"))
```

You will notice that the file has distorted sound. I looked at the wavform, it is actually not clipping. There's something more fishy going on in the conversion to double. I've looked at the code there and noticed some issues but wasn't able to fix the issue. I'm starting a PR in `ocaml-ffmpeg` with what I have noticed. See ya there!

**Comment by gndl** [comment]
Sorry, I forgot to mention in the documentation that the data returned by Swresample.convert is invalidated by a new call to this function. The following patch solve the problem.

diff --git a/src/decoder/ffmpeg_decoder.ml b/src/decoder/ffmpeg_decoder.ml
index d5c9ba64..f8d67794 100644
--- a/src/decoder/ffmpeg_decoder.ml
+++ b/src/decoder/ffmpeg_decoder.ml
@@ -100,7 +100,7 @@ let create_decoder fname =
       Frame.master_of_audio (Array.length data.(0))
     in
     decr_remaining consumed;
-    data
+    Array.init (Array.length data) (fun i -> Array.copy data.(i))
   in
   let seek ticks =
     let position = Frame.seconds_of_master ticks in



Once the conversion is complete, liquidsoap enters an infinite loop with CPU usage of 98%. I do not know if it's normal or if it's a problem with ffmpeg.

**Comment by gndl** [comment]
To avoid this problem, I will make optional the reuse of the output buffer with the release of the buffer as the default behavior. I hesitate to remove the possibility of using this optimization because my bench showed a performance gain of more than 2 percent.

**Comment by toots** [comment]
Ok, trying it. Re-using output buffer is quite okay for me as long as it's properly documented.

**Comment by toots** [comment]
Yup, that works thanks. Will close the other issue. I've seen a couple of segfaults though, will report if I can get a stack trace.

**Comment by LemonBoy** [comment]
> I am not really concerned by the video part so i didn't do the video frame/planes (bigarray) conversion functions.

Any update about this? I'd love to use `ocaml-ffmpeg` to do some processing on the frames and indeed `Avutil.Video.frame_get` is way too slow.

**Comment by gndl** [comment]
Yes, I recently merged the swscale_converter branch in the master branch.
You can see a (limited) use case [here](https://github.com/savonet/ocaml-ffmpeg/blob/master/examples/demuxing_decoding/demuxing_decoding.ml).
Be careful, exept from the example above, this code has not been tested.

**Comment by LemonBoy** [comment]
Sure but I'd like to skip the resampling process. My idea was to have some kind of thin wrapper around the `AVframe` contents.

**Comment by gndl** [comment]
@LemonBoy, In the last commit, I experimented with several ways to access a video frame. The test result can be obtained by running the examples/encode_video example like this:
`make test`
For reading the data:
- The Bytes_copy.get_image function copies video frame data into byte arrays.
- The Bigarray_copy.get_image function copies video frame data into Bigarrays.
- The Set.get_image function directly accesses the video frame data with x and y coordinates
- The Planar_set.get_image function directly accesses the video frame data in a planar manner.
- The Wrap.get_image function wraps the video frame data in Bigarrays without copying it and uses the Bigarray accessors.

For writing data (fill_image), in addition to previously used methods, copying data to the video frame using the Swscale module is also tested.
 
On my (old) machine, I get the following results:

Bigarray_copy.get_image 0                : 1.244290 seconds
Bytes_copy.get_image 0                   : 1.320332 seconds
Planar_set.get_image 0                   : 5.049537 seconds
Set.get_image 0                          : 6.395155 seconds
Wrap.get_image 0                         : 8.161680 seconds

Bytes_copy.fill_image                    : 2.905994 seconds
Bigarray_copy.fill_image                 : 3.622380 seconds
Bytes_scale.fill_image                   : 3.660806 seconds
Bigarray_scale.fill_image                : 4.024561 seconds
Planar_set.fill_image                    : 7.562174 seconds
Set.fill_image                           : 8.476870 seconds
Wrap.fill_image                          : 11.416126 seconds

For reading, it is the copy to Bigarray that is the most effective. For writing, it is the copy from bytes arrays that is the most efficient (which is annoying because Bigarrays are more convenient for manipulating integers).
The wrapping method is particularly slow because, in order to avoid access after the video frame has been garbage-collected, it is necessary to encapsulate the structure.
@toots, @LemonBoy, tell me which method is best for you, I will keep this one.
All your ideas are welcome.

**Comment by LemonBoy** [comment]
On my reasonably new laptop I get the following results, I had to remove `-ffast-math` because that's not supported by ocamlopt (the i386 backend is the only one supporting it).

<details/>
  <summary>OCaml 4.06</summary>

```
Bigarray_copy.get_image 31188320256      : 1.026108 seconds
Bytes_copy.get_image 31188320256         : 1.037922 seconds
Planar_set.get_image 31188320256         : 2.213529 seconds
Set.get_image 31188320256                : 2.528611 seconds
Wrap.get_image 31188320256               : 3.659002 seconds

Bigarray_copy.fill_image                 : 0.775497 seconds
Bigarray_scale.fill_image                : 0.938345 seconds
Bytes_copy.fill_image                    : 1.570437 seconds
Bytes_scale.fill_image                   : 1.738604 seconds
Planar_set.fill_image                    : 2.664477 seconds
Set.fill_image                           : 2.982687 seconds
Wrap.fill_image                          : 5.537404 seconds
```

</details>

<details>
<summary>OCaml 4.06+flambda</summary>

```
Bytes_copy.get_image 31188320256         : 1.104021 seconds
Bigarray_copy.get_image 31188320256      : 1.130194 seconds
Planar_set.get_image 31188320256         : 2.360727 seconds
Set.get_image 31188320256                : 2.527284 seconds
Wrap.get_image 31188320256               : 3.767551 seconds

Bigarray_copy.fill_image                 : 0.886269 seconds
Bytes_copy.fill_image                    : 0.967497 seconds
Bigarray_scale.fill_image                : 1.046061 seconds
Bytes_scale.fill_image                   : 1.146593 seconds
Planar_set.fill_image                    : 2.839119 seconds
Set.fill_image                           : 3.072064 seconds
Wrap.fill_image                          : 5.695171 seconds
```

</details>

Sadly flambda gives no advantage here, I'd go by the `Bigarray` route given the results.

> The wrapping method is particularly slow because, in order to avoid access after the video frame has been garbage-collected, it is necessary to encapsulate the structure.

That's surprising, I didn't expect the per-access overhead to be so big even though there's a pointer indirection in between.

**Comment by gndl** [comment]
Thank you for your feedback.

> That's surprising, I didn't expect the per-access overhead to be so big even though there's a pointer indirection in between.

The user must access the bigarray by a function call. I had hoped that this function would be inlined but it seems that it is not the case.

**Comment by gndl** [comment]
@LemonBoy, I think I have found a way to use a frame wrapper with acceptable efficiency and safety.
The user passes a callback function that receives the planes as parameter. Access to the frame through the bigarray is safe as long as it occurs in the callback function and the frame is not sent to an encoder.
I get the following results:
```
Wrap.get_image 0                         : 0.962648 seconds
Bigarray_copy.get_image 0                : 1.248070 seconds
Bytes_copy.get_image 0                   : 1.547825 seconds
Planar_set.get_image 0                   : 3.935357 seconds
Set.get_image 0                          : 6.432126 seconds

Wrap.fill_image                          : 1.767915 seconds
Bytes_copy.fill_image                    : 2.939943 seconds
Bigarray_copy.fill_image                 : 3.650611 seconds
Bytes_scale.fill_image                   : 3.686377 seconds
Bigarray_scale.fill_image                : 4.123058 seconds
Planar_set.fill_image                    : 4.710555 seconds
Set.fill_image                           : 8.511368 seconds
````
Hoping that it meets your needs.

**Comment by LemonBoy** [comment]
That's indeed much faster, thanks for the quick update!

**Comment by gndl** [comment]
@toots, how can I reproduce the segfaults ?

**Comment by toots** [comment]
They used to happen when using the encoder with liquidsoap. I believe it's the branch at https://github.com/savonet/liquidsoap/pull/547/files. I have to carve time to try the latest `ocaml-ffmpeg` again.

**Comment by kit-ty-kate** [comment]
What is the status of this? I'd be quite interested to be able to release some software relying on the new API (I'm only using `FFmpeg.Av.open_input`, `FFmpeg.Av.open_input` and `FFmpeg.Av.close` but still ^^)

**Comment by gndl** [comment]
I think the library is almost ready to be delivered. I just want to correct the malfunction found with liquidsoap before delivering. I have not had the time to deal with it in recent months but I intend to do it these days.

**Comment by toots** [comment]
Cool. I'm currently focusing on releasing liquidsoap so I'll have another pass at it very soon.

**Comment by toots** [comment]
@gndl Yeah, I'm sorry to report that I'm still seeing a lot of random crashes that seem to be associated with stack corruption.

The liquidsoap code is threaded, have you tested your binding in a multithreaded context?

If you want to try our code, you can use the `ffmpeg-decoder` branch. I'll be happy to provide more instructions if you wish to go that route.

**Comment by toots** [comment]
You can simply enable `ao` and do:
```
liquidsoap 'output.ao(fallible=true,single("/path/to/file.mp3"))'
```
I see segfaults with this script. Make sure that `mad` decoder is disabled.

**Comment by gndl** [comment]
Maybe I can not reproduce the crash because my CPU is mono core. Any way, I added a lock manager for the thread unsafe functions with the "av_lockmgr_register" API. Note that as of version 4.0, this API is deprecated and ffmpeg must be build with thread support to get basic thread safety.
Let me know if it solve the problem.

**Comment by toots** [comment]
Here's one:
```
* thread #9, stop reason = signal SIGABRT
  * frame #0: 0x00007fff54ccdb6e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff54e98080 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff54c291ae libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff54d27822 libsystem_malloc.dylib`free + 521
    frame #4: 0x00000001008a8660 libavutil.55.dylib`___lldb_unnamed_symbol7$$libavutil.55.dylib + 58
    frame #5: 0x00000001008a83bf libavutil.55.dylib`___lldb_unnamed_symbol6$$libavutil.55.dylib + 81
    frame #6: 0x00000001008aeeb9 libavutil.55.dylib`av_frame_unref + 42
    frame #7: 0x00000001008aee84 libavutil.55.dylib`av_frame_free + 22
    frame #8: 0x000000010025122e liquidsoap`finalize_frame(v=<unavailable>) at avutil_stubs.c:218 [opt]
    frame #9: 0x0000000100283978 liquidsoap`caml_empty_minor_heap at minor_gc.c:381 [opt]
    frame #10: 0x0000000100284045 liquidsoap`caml_gc_dispatch at minor_gc.c:438 [opt]
    frame #11: 0x0000000100284194 liquidsoap`caml_check_urgent_gc(extra_root=<unavailable>) at minor_gc.c:482 [opt]
    frame #12: 0x000000010025bd4d liquidsoap`convert_to_planar_float_array(swr=0x0000000102908970, in_nb_samples=1152, out_nb_samples=<unavailable>) at swresample_stubs.c:361 [opt]
    frame #13: 0x000000010025a572 liquidsoap`ocaml_swresample_convert(_swr=4300784960, _in_vector=4379154864) at swresample_stubs.c:466 [opt]
```
Segfault happens during a GC memory collection and seems linked to a thread cleanup. It's on OSX so it could be a ffmpeg bug on OSX.

**Comment by toots** [comment]
Well the `ffmpeg` code seems a bit complex but my best bet would be that the GC is freeing a frame that is still being used somewhere else.

**Comment by gndl** [comment]
It sounds like a double free like the one described [here](https://github.com/aubio/aubio/issues/103).
Can you upgrade FFmpeg to the latest version (4.0.2) in order to see if the problem still exists ?

**Comment by toots** [comment]
I'm seeing a different error:
```
* thread #2, stop reason = EXC_BAD_ACCESS (code=1, address=0xc)
  * frame #0: 0x00000001007de1a4 libavutil.56.dylib`av_buffer_ref + 44
    frame #1: 0x0000000100afe9bb libavcodec.58.dylib`av_packet_ref + 34
    frame #2: 0x0000000100b6422d libavcodec.58.dylib`avcodec_send_packet + 148
    frame #3: 0x000000010021a7e0 liquidsoap`decode_packet(av=<unavailable>, stream=<unavailable>, packet=<unavailable>, frame=<unavailable>) at av_stubs.c:554 [opt]
    frame #4: 0x000000010021a6df liquidsoap`ocaml_av_read_stream_frame(_stream=4334573384) at av_stubs.c:686 [opt]
    frame #5: 0x000000010002910d liquidsoap`camlFfmpeg_decoder__decode_2625 + 29
    frame #6: 0x0000000100109a28 liquidsoap`camlDecoder__fill_102011 + 248
    frame #7: 0x00000001000bfef4 liquidsoap`camlRequest_source__fun_603163 + 20
    frame #8: 0x000000010011fa3c liquidsoap`camlTutils__mutexify_1007 + 108
    frame #9: 0x000000010011fa3c liquidsoap`camlTutils__mutexify_1007 + 108
    frame #10: 0x00000001000bdb0b liquidsoap`camlRequest_source__fun_603193 + 235
    frame #11: 0x00000001000be336 liquidsoap`camlRequest_source__fun_603333 + 22
    frame #12: 0x00000001000fdd8e liquidsoap`camlSource__fun_3160 + 926
    frame #13: 0x0000000100095ea8 liquidsoap`camlOutput__fun_2830 + 344
    frame #14: 0x00000001000fb13c liquidsoap`camlClock__fun_1003315 + 444
    frame #15: 0x00000001001c1818 liquidsoap`camlStdlib__list__fold_left_1165 + 40
    frame #16: 0x00000001000f9611 liquidsoap`camlClock__fun_1003268 + 225
    frame #17: 0x00000001000fc5c1 liquidsoap`camlClock__loop_702568 + 785
    frame #18: 0x00000001000f9d6d liquidsoap`camlClock__fun_1003512 + 749
    frame #19: 0x00000001001201f2 liquidsoap`camlTutils__fun_902702 + 802
    frame #20: 0x00000001001b3d39 liquidsoap`camlThread__fun_2005 + 137
    frame #21: 0x000000010025d964 liquidsoap`caml_start_program + 92
    frame #22: 0x0000000100233688 liquidsoap`caml_thread_start + 104
    frame #23: 0x00007fff54e95661 libsystem_pthread.dylib`_pthread_body + 340
    frame #24: 0x00007fff54e9550d libsystem_pthread.dylib`_pthread_start + 377
    frame #25: 0x00007fff54e94bf9 libsystem_pthread.dylib`thread_start + 13
```

Considering that the code of the binding has a lot of global ocaml vars, I'd suggest looking this direction. This all seems like a GC problem to me.

**Comment by gndl** [comment]
The last commit removes Av module's use of global ocaml vars in the liquidsoap use case.
The last commit removes any use of global ocaml vars by the Av module in the liquidsoap use case.
It will take me a little more time to do the same for the Swresample module. Feel free to test now or wait for the total removal of global ocaml vars.
If this solves the problem, I will remove them permanently.

**Comment by toots** [comment]
You might be in the good direction. Here's what I see now:
```
* thread #9, stop reason = signal SIGABRT
  * frame #0: 0x00007fff59496b6e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff59661080 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff593f21ae libsystem_c.dylib`abort + 127
    frame #3: 0x0000000101d2688d libswresample.3.dylib`___lldb_unnamed_symbol7$$libswresample.3.dylib + 617
    frame #4: 0x0000000101d2f748 libswresample.3.dylib`___lldb_unnamed_symbol101$$libswresample.3.dylib + 71
    frame #5: 0x0000000101d2f5f2 libswresample.3.dylib`swr_convert + 903
    frame #6: 0x000000010021f737 liquidsoap`convert_to_planar_float_array(swr=0x00000001023008e0, in_nb_samples=47, out_nb_samples=47) at swresample_stubs.c:351 [opt]
    frame #7: 0x000000010021dfa2 liquidsoap`ocaml_swresample_convert(_swr=4300200664, _in_vector=4347320560) at swresample_stubs.c:466 [opt]
    frame #8: 0x0000000100028efc liquidsoap`camlFfmpeg_decoder__convert_2571 + 28
```

**Comment by gndl** [comment]
Ok, the last commit removes Swresample module's use of global ocaml vars. Goodbye optimization.

**Comment by toots** [comment]
Unfortunately, I'm sorry to report:
```
* thread #9, stop reason = signal SIGABRT
  * frame #0: 0x00007fff59496b6e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff59661080 libsystem_pthread.dylib`pthread_kill + 333
    frame #2: 0x00007fff593f21ae libsystem_c.dylib`abort + 127
    frame #3: 0x0000000101d0ea67 libswresample.3.dylib`___lldb_unnamed_symbol7$$libswresample.3.dylib + 626
    frame #4: 0x0000000101d177cb libswresample.3.dylib`___lldb_unnamed_symbol101$$libswresample.3.dylib + 74
    frame #5: 0x0000000101d175fa libswresample.3.dylib`swr_convert + 903
    frame #6: 0x0000000100213cfb liquidsoap`convert_to_planar_float_array(swr=0x0000000103f0bf20, in_nb_samples=<unavailable>, out_nb_samples=<unavailable>, out_vector=0x0000700005ddb9f8) at swresample_stubs.c:337 [opt]
    frame #7: 0x0000000100212620 liquidsoap`ocaml_swresample_convert(_swr=4300070168, _in_vector=4355718344) at swresample_stubs.c:446 [opt]
```

Still looks like an issue with memory management to me. If you care to explain the code a little bit more I could look at it further.

**Comment by gndl** [comment]
This may not be the end of our problems but this time I think I have fixed a serious bug. I was misusing the av_samples_alloc function for planar data. I do not understand why it did not crash under Linux.
I will try to explain the code in general but if there are points that you would like to see clarified, do not hesitate to indicate them to me.

**Comment by toots** [comment]
Well, it's looking good here! Been playing for 20 min without issues..

**Comment by toots** [comment]
Yeah, it's looking good. I've been playing for hours without issues with the global variables in. As far as I'm concerned the bug is fixed 😎Just pushed the revert of the two commits removing the globals. Nice work!

**Comment by gndl** [comment]
I get a segfault when I use logging with older version of OCaml (found on versions 4.02.3 to 4.04.2).
Here is the call stack:
#0  0x00421de6 in caml_execute_signal ()
#1  0x00421ea7 in caml_process_pending_signals ()
#2  0x00421f05 in caml_leave_blocking_section ()
#3  0x0040c07b in ocaml_avcodec_send_frame (_ctx=<optimized out>, _frame=<optimized out>) at avcodec_stubs.c:592
#4  0x00432f8c in caml_interprete ()
#5  0x0041e171 in caml_main ()
#6  0x00431995 in main ()

The problem disappears if I comment the ocaml_ffmpeg_register_thread call in the av_log_ocaml_callback function. Could this come from the fact that the thread would be registered twice?

**Comment by gndl** [comment]
Answer to myself :
Registering an already-register thread does nothing and in my case, registration is OK (return 1).

**Comment by toots** [comment]
Liquidsoap will ship with support for the new `ffmpeg` API shortly.

**Comment by gndl** [comment]
I was planning to add the "available" field to the ffmpeg.opam file with ocaml-version >= 4.02.3. Should we start from version 4.05.0 because of the C call to OCaml problem ?
In addition, should the license statement be added to the source files?

**Comment by toots** [comment]
Considering how fast OCaml is moving these days I don't think it should be an issue to make available only post `4.05.0`. That's also the current version in Debian testing. You may or may not add the license statement, it's up to you. When I was packaging with Debian we considered the main `COPYRIGHT` as the default license when no headers were provided in the source code.

**Comment by gndl** [comment]
OK, I added OCaml version 4.05.0 as a prerequisite in the opam file.
Since this is not essential, I will not add the license header in the source files.
I am delighted to see the release of ocaml-ffmpeg :-).

**Comment by toots** [comment]
I'm releasing liquidsoap imminently, I'll also do the `ffmpeg` release if that's cool with you.
